cni/pkg/ebpf/server/redirectionServer.go

```diff
@@ -0,0 +1,592 @@
+	TcPrioFilter         = 1      // refer to include/uapi/linux/pkt_sched.h TC_PRIO_FILLER
+)
+
+var isBigEndian = native.IsBigEndian
+
+type RedirectServer struct {
+	redirectArgsChan           chan *RedirectArgs
+	obj                        ambient_redirectObjects
+	ztunnelHostingressFd       uint32
+	ztunnelHostingressProgName string
+	ztunnelIngressFd           uint32
+	ztunnelIngressProgName     string
+	inboundFd                  uint32
+	inboundProgName            string
+	outboundFd                 uint32
+	outboundProgName           string
+}
+
+func (r *RedirectServer) initBpfObjects() error {
+	var options ebpf.CollectionOptions
+	if _, err := os.Stat(MapsPinpath); err != nil {
+		if os.IsNotExist(err) {
+			if err := os.Mkdir(MapsPinpath, os.ModePerm); err != nil {
+				return fmt.Errorf("unable to create ambient bpf mount directory: %v", err)
+			}
+		}
+	}
+	options.Maps.PinPath = MapsPinpath
+	// load ebpf program
+	obj := ambient_redirectObjects{}
+	if err := loadAmbient_redirectObjects(&obj, &options); err != nil {
+		return fmt.Errorf("loading objects: %v", err)
+	}
+	r.obj = obj
+	r.ztunnelHostingressFd = uint32(r.obj.ZtunnelHostIngress.FD())
+	ztunnelHostingressInfo, err := r.obj.ZtunnelHostIngress.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.ztunnelHostingressProgName = ztunnelHostingressInfo.Name
+	r.ztunnelIngressFd = uint32(r.obj.ZtunnelIngress.FD())
+	ztunnelIngressInfo, err := r.obj.ZtunnelIngress.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.ztunnelIngressProgName = ztunnelIngressInfo.Name
+
+	r.inboundFd = uint32(r.obj.AppInbound.FD())
+	inboundInfo, err := r.obj.AppInbound.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.inboundProgName = inboundInfo.Name
+	r.outboundFd = uint32(r.obj.AppOutbound.FD())
+	outboundInfo, err := r.obj.AppOutbound.Info()
+	if err != nil {
+		return fmt.Errorf("unable to load metadata of bfp prog: %v", err)
+	}
+	r.outboundProgName = outboundInfo.Name
+	return nil
+}
+
+// Note: this struct should be exactly the same defined in C
+// it will be encoded byte by byte into memory
+type mapInfo struct {
+	Ifindex uint32
+	MacAddr [6]byte
+	Flag    uint8
+	Pad     uint8
+}
+
+func NewRedirectServer() *RedirectServer {
+	if err := checkOrMountBPFFSDefault(); err != nil {
+		log.Fatalf("BPF filesystem mounting on /sys/fs/bpf failed: %v", err)
+	}
+
+	if err := setLimit(); err != nil {
+		log.Fatalf("Setting limit failed: %v", err)
+	}
+
+	r := &RedirectServer{
+		redirectArgsChan: make(chan *RedirectArgs),
+	}
+
+	if err := r.initBpfObjects(); err != nil {
+		log.Fatalf("Init bpf objects failed: %v", err)
+	}
+
+	return r
+}
+
+func checkOrMountBPFFSDefault() error {
+	var err error
+
+	_, err = os.Stat(MapsRoot)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(MapsRoot, 0o755); err != nil {
```

Comment:
1. Are there any legit scenarios where `/sys/fs/bpf` not being present is something we should try to fix in this way?

2. Should we try to fix this by `mkdir`-ing in sysfs ourselves? Or should we bail, as an indication that bpf isn't supported by this kernel?